src/multiplex.c

```diff
@@ -183,6 +186,25 @@ static _Noreturn void *process_out(struct thread_args *args) {
     struct header hdr;
     sigset_t empty_set;
 
+    assert(read_fds_len <= MAX_FDS);
+    /*
+     * We must not block reading from file descriptors in this function
+     * because if we do, a deadlock is possible.  For instance, gpg
+     * could have written a few bytes to this file descriptor, and now
+     * be waiting for us to read a large number of bytes from another
+     * file descriptor.  If we try to read more bytes than are available,
+     * and the file descriptor is in blocking mode, deadlock can result.
```

Comment:
No, `read()` will return the data immediately if there is some data available. The only case where blocking/non-blocking mode matters is when no data is available at all. And in this particular case, we know some data is available, because `pselect()` reported so.

The freeze you are observing is probably something else. I would rather guess it's Thunderbird not reading data from split-gpg, while writing to it.

Anyway, probably not worth investing too much time in debugging nor fixing it, because splig-gpg2.

Reply:
Fix was easy (#80).